packages/aws-cdk-lib/core/lib/stage.ts

```diff
@@ -210,13 +217,52 @@ export class Stage extends Construct {
    * calls will return the same assembly.
    */
   public synth(options: StageSynthesisOptions = { }): cxapi.CloudAssembly {
-    if (!this.assembly || options.force) {
+
+    const newConstructPaths = new Set<string>();
+    recurseAndlistAllConstructPaths(this);
+
+    // If the assembly cache is uninitiazed, run synthesize and re-run list all construvt paths
+    if (this.constructPathsCache.size == 0 || !this.assembly || options.force) {
       this.assembly = synthesize(this, {
         skipValidation: options.skipValidation,
         validateOnSynthesis: options.validateOnSynthesis,
       });
+      recurseAndlistAllConstructPaths(this);
+      this.constructPathsCache = newConstructPaths;
+    }
+
+    // Lists all construct paths
+    function recurseAndlistAllConstructPaths(construct: IConstruct) {
+      newConstructPaths.add(construct.node.path);
+      for (const child of construct.node.children) {
+        if (!Stage.isStage(child)) {
+          recurseAndlistAllConstructPaths(child);
+        }
+      }
+    }
+
+    // If the construct paths set has changed
+    if (!constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      if (options.warnInsteadOfError) {
+        Annotations.of(this).addWarningV2('@aws-cdk/core:synth-warning', 'Construct paths have been modified since the last synth call.');
```

Comment:
Similar to the error, I don't really know what this warning is telling me and why it is a warning.